### PR TITLE
Fix restarting a websocket after a network error

### PR DESCRIPTION
**Pull request by martinbonnin**
This PR serializes all messages from the `WsProtocol` in the "supervisor". 

What I call "supervisor" is a long-lived coroutine that:
- receives `Commands` from the various subscription `Flow`s, creates a `WsProtocol` when needed and forwards the commands
- receives `Events` from the `WsProtocol` and forwards them back to the various subscription `Flow`s

The `WsProtocol` owns the `WebSocketConnection`. The "supervisor" owns the `WsProtocol` and closes it after `idleTimeout`. 

The problem this PR adresses is that there was no way previously to know if a `WsProtocol` ended as an error and therefore was never re-created. With these changes, `WsProtocol.run` is required to send a terminal `NetworkError` event that will be received by the "supervisor" that can then reset its state.

Fixes https://github.com/apollographql/apollo-android/issues/3634

### CODE REVIEW

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt [hunk 1]
```diff
@@ -77,74 +74,117 @@ private constructor(
...
+            try {
+              protocol.connectionInit()
+            } catch (e: Exception) {
+              // Error initializing the connection
+              protocol = null
+              mutableEvents.emit(NetworkError(e))
+              continue
+            }
+
+            connectionJob = scope.launch(start = CoroutineStart.UNDISPATCHED) {
```
Comment (BoD) [inline apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:163]: Question: why do de need `UNDISPATCHED` here?

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/internal/WsMessage.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
+package com.apollographql.apollo3.network.ws.internal
+
+import com.apollographql.apollo3.api.ApolloRequest
+import com.apollographql.apollo3.api.Operation
+import kotlinx.coroutines.channels.Channel
+
+internal class ConnectionHolder(val events: Channel<Event>) {
```
Comment (BoD) [inline apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/internal/WsMessage.kt:7]: Is this unused?

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:163]: I wanted to make sure `protocol.run()` would always be called. I'm not 100% sure if `protocol` could be reset before starting to execute the coroutine so added this as an extra precaution. Maybe it's not required but it shouldn't hurt so better safe than sorry...

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/internal/WsMessage.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/internal/WsMessage.kt:7]: Good catch ! 🙏

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt (hunk 1 above)
Comment (BoD) [reply apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:163]: 👍 Makes sense, thanks (and TIL 😄)!

Comment (martinbonnin) [reply apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:163]: Side note: Looks like the type inference is not working here. `protocol` should be nullable always. I filed https://youtrack.jetbrains.com/issue/KT-50092

**Review by BoD — APPROVED** [review]

**Pull request merged**